### ISSUE TITLE
POST /chaincode deploy does not accept request object id of type string 

### ISSUE THREAD

**Issue by jjjjibm** [hyperledger-archives/fabric#1100]
I am in the process of converting the sdk from using the deprecated devops API to using the chaincode API.   Here is a description of the new API: https://github.com/hyperledger/fabric/blob/master/docs/API/CoreAPI.md#chaincode

Inspecting the deploy request format, we see a new field here called ID.  The samples in the API use integers as sample values.  The RPC specification, nowever http://www.jsonrpc.org/specification#request_object , says the ID can also be a string:

> id   An identifier established by the Client that MUST contain a String, Number, or NULL value if included. 

I tried setting the ID to the string "deploy1": 

{
    "jsonrpc": "2.0",
    "method": "deploy",
    "params": {
        "type": 1,
        "chaincodeID": {
            "path": "github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02"
        },
        "ctorMsg": {
            "function": "init",
            "args": [
                "a",
                "20000",
                "b",
                "9000"
            ]
        },
        "secureContext": "diego"
    },
    "id": "deploy1"
}

but the API rejected this value:

{
    "jsonrpc": "2.0",
    "error": {
        "code": -32700,
        "message": "Parse error",
        "data": "Error unmarshalling chaincode request payload: json: cannot unmarshal string into Go value of type int64"
    },
    "id": null
}

Changing the "id" from "deploy1" to numeric one enabled the REST call to work: 
{
    "jsonrpc": "2.0",
    "method": "deploy",
    "params": {
        "type": 1,
        "chaincodeID": {
            "path": "github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02"
        },
        "ctorMsg": {
            "function": "init",
            "args": [
                "a",
                "20000",
                "b",
                "9000"
            ]
        },
        "secureContext": "diego"
    },
    "id": 1
}

**Sending Rest Request to : http://127.0.0.1:3000/chaincode/

Response from Rest Call: >>

{
    "jsonrpc": "2.0",
    "result": {
        "status": "OK",
        "message": "401bed727b2e662dcf38e331636db2dee24535607da5d78ed3d9dc0bd9712d4b0fc7552233cf65ef2e92951f7f7b2a3b8671954fc885ac934bca67bc9703eb0e"
    },
    "id": 1
}

Either the specification is wrong, or our implementation of the /chaincode REST API call is wrong.

**Comment by kletkeman** [comment]
Good catch. The schema in rest_api.json defines the "id" property as a number:

```
          "id": {
             "type": "integer",
             "format": "int64",
             "description": "An integer number used to correlate the request and response objects. If it is not included, the request is assumed to be a notification and the server will not generate a response."
          }
```

But the spec does allow integer (JSON number with no fractional part), string and null. Here is a sample schema that models it to match:

https://github.com/fge/sample-json-schemas/blob/master/jsonrpc2.0/jsonrpc-request-2.0.json

```
            "id": {
                "type": [ "string", "number", "null" ],
                "note": [
                    "While allowed, null should be avoided: http://www.jsonrpc.org/specification#id1",
                    "While allowed, a number with a fractional part should be avoided: http://www.jsonrpc.org/specification#id2"
                ]
            },
```

**Comment by kletkeman** [comment]
Also note that, if null is sent as of this comment's time and date, hyperledger peer returns nothing at all for deploy or query (and presumably invoke although I have not confirmed that.)

**Comment by angrbrd** [comment]
This issue has been addressed by PR #1142.
